include/graaflib/algorithm/topological_sorting.tpp

```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include <graaflib/algorithm/cycle_detection.h>
+#include <graaflib/algorithm/topological_sorting.h>
+
+#include <vector>
+#include <unordered_map>
+
+namespace graaf::algorithm {
+
+namespace detail {
+
+enum class vertex_status { UNVISITED, VISITED};
+
+// DFS topological sort
+template <typename V, typename E>
+void do_dfs_topological_sort(
+    const graph<V, E, graph_type::DIRECTED>& graph,
+    vertex_id_t start_vertex,
+    std::unordered_map<vertex_id_t, vertex_status>& processed_vertices,
+    std::vector<vertex_id_t>& sorted_vertices) {
+    
+    processed_vertices[start_vertex] = vertex_status::VISITED;
+    for (const auto& next_vertex : graph.get_neighbors(start_vertex)) {
+        if (processed_vertices[next_vertex] == vertex_status::UNVISITED) {
+            do_dfs_topological_sort(graph, next_vertex, processed_vertices,
+                sorted_vertices);
+        }
+    }
+
+        sorted_vertices.push_back(start_vertex);
+    }
+
+}; // namespace detail
+
+template <typename V, typename E>
+std::vector<vertex_id_t> topological_sort(
+    const graph<V, E, graph_type::DIRECTED>& graph) {
+    
+    // Graph should be acyclic
+    if (dfs_cycle_detection(graph)) {
+        return {};    
+    }
+
+    using enum detail::vertex_status;
+    std::vector<vertex_id_t> sorted_vertices{};
+    sorted_vertices.reserve(graph.vertex_count());
+    std::unordered_map<vertex_id_t, detail::vertex_status> processed_vertices{};
+
+    for (const auto& vertex : graph.get_vertices()) {
+        if (processed_vertices[vertex.first] == UNVISITED) {
+            detail::do_dfs_topological_sort(graph, vertex.first, processed_vertices, sorted_vertices);
```

Comment:
[very minor]: To make this slightly more readable, how about some structured binding here to unpack the pair:
```suggestion
    for (const auto& [vertex_id, _] : graph.get_vertices()) {
        if (processed_vertices[vertex.first] == UNVISITED) {
            detail::do_dfs_topological_sort(graph, vertex_id, processed_vertices, sorted_vertices);
```

Reply:
Oh, good advice. I guess I will correct my other PR. I wasn't satisfied with 'vertex.first' aswell!